eclipse-che/che-operator
# Allow enabling persistence of /home/user/ in workspaces

### What does this PR do?
This PR adds 2 new fields to the Che Cluster CR:
-  `spec.devEnvironments.persistUserHome`: holds config settings related to the persistence of `/home/user/` in workspaces 
- `spec.devEnvironments.persistUserHome.enabled` determines whether `/home/user/` will persist in workspaces. Disabled by default.

Note: this PR depends on DWO 0.22.0 which has currently only been released upstream.



### Screenshot/screencast of this PR
n/a

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/22102

### How to test this PR?

Ensure that changes to `spec.devEnvironments.persistUserHome.enabled` from the Che Cluster CR are propagated to the Che-owned DWOC's `config.workspace.persistUserHome`:

1. Setting `spec.devEnvironments.persistUserHome.enabled: true` in the Che Cluster CR results in `config.workspace.persistUserHome.enabled: true` for the Che-owned DWOC
2. Setting `spec.devEnvironments.persistUserHome.enabled: false` in the Che Cluster CR results in `config.workspace.persistUserHome.enabled: false` for the Che-owned DWOC
3. Removing `spec.devEnvironments.persistUserHome` in the Che Cluster CR (i.e. setting it to nil) removes `config.workspace.persistUserHome` in the Che-owned DWOC

For example, for test case 1. do:
1. `kubectl edit checluster -n eclipse-che`
```diff
(...)
spec:
(...)
  devEnvironments:
    containerBuildConfiguration:
      openShiftSecurityContextConstraint: container-build
    defaultNamespace:
      autoProvision: true
      template: <username>-che
    maxNumberOfWorkspacesPerUser: -1
+    persistUserHome:
+      enabled: true
    secondsOfInactivityBeforeIdling: 1800
    secondsOfRunBeforeIdling: -1
    startTimeoutSeconds: 300
    storage:
      pvcStrategy: per-user
```
2. `kubectl describe dwoc -n eclipse-che` and ensure `persistUserHome` is now enabled:
```diff
Name:         devworkspace-config  
Namespace:    eclipse-che  
Labels:       <none>  
Annotations:  <none>  
API Version:  controller.devfile.io/v1alpha1  
Config:  
 Workspace:  
   Container Security Context:  
     Allow Privilege Escalation:  true  
     Capabilities:  
       Add:  
         SETGID  
         SETUID  
   Deployment Strategy:  Recreate  
+   Persist User Home:  
+     Enabled:         true  
   Progress Timeout:  300s  
   Service Account:  
     Disable Creation:  false  
Kind:                    DevWorkspaceOperatorConfig
```

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.
